apps/prairielearn/src/question-servers/freeform.js

```diff
@@ -1816,25 +1878,35 @@ export async function grade(submission, variant, question, question_course) {
   });
 }
 
+/**
+ * @param {import('../lib/db-types.js').Variant} variant
+ * @param {import('../lib/db-types.js').Question} question
+ * @param {import('../lib/db-types.js').Course} course
+ * @param {'correct' | 'incorrect' | 'invalid'} test_type
+ * @returns {import('./types.js').QuestionServerReturnValue<import('./types.js').TestResultData>}
+ */
 export async function test(variant, question, course, test_type) {
   return instrumented('freeform.test', async () => {
     debug('test()');
     if (variant.broken_at) throw new Error('attempted to test broken variant');
 
     const context = await getContext(question, course);
+
+    /** @satisfies {import('./types.js').ExecutionData & { test_type: 'correct' | 'incorrect' | 'invalid' }} */
     let data = {
-      params: variant.params,
-      correct_answers: variant.true_answer,
+      params: variant.params ?? {},
+      correct_answers: variant.true_answer ?? {},
```

Comment:
As above, these should never be nullish in practice, but that's not encoded in the database schema.

Reply:
Is this worth a comment in the code?

Reply:
Comments added in a75b7d1.